internal/fetch/fetch.go

```diff
@@ -22,14 +25,25 @@ func Fetch(ctx context.Context, hash, target, folder string, insecure bool) erro
 	if err != nil {
 		return err
 	}
-	newTarget := strings.TrimPrefix(target, u.Scheme+"://")
+	source := strings.TrimPrefix(target, u.Scheme+"://")
 	switch u.Scheme {
 	case "oci":
-		return oci.Fetch(ctx, hash, newTarget, folder, insecure)
+		return oci.Fetch(ctx, hash, source, folder, insecure)
 	case "s3":
-		return fmt.Errorf("s3 schema not (yet) supported: %s", target)
+		source = strings.ReplaceAll(source, "{{hash}}", hash)
+		file, _ := os.CreateTemp("", "cache.tar")
+		if err := provider.Providers[u.Scheme].Fetch(ctx, source, file.Name()); err != nil {
+			return err
+		}
+		f, err := os.Open(file.Name())
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		return extract.Archive(ctx, f, folder, nil)
```

Comment:
I think this one should look as simple as the other one, like `s3.Fetch` (and "hide" implementation detail of the tarball, …)

Reply:
I am trying to keep a common implementation for all providers.
Once we do it for S3. we can remove all other conditions and  provider.Providers[u.Scheme] will get the correct provider where we can call update.

and creating/extracting tarball is common task for all providers so I wanted to have that here only

Reply:
final implementation should look like  this

`func Upload(ctx context.Context, hash, target, folder string, insecure bool) error {
	u, err := url.Parse(target)
	if err != nil {
		return err
	}
	newTarget := strings.TrimPrefix(target, u.Scheme+"://")
	newTarget = strings.ReplaceAll(newTarget, "{{hash}}", hash)
	tarFile, err := os.CreateTemp("", "cache.tar")
	if err != nil {
		return err
	}
	if err := fileutils.Tarit(folder, tarFile.Name()); err != nil {
		return err
	}
	defer os.Remove(tarFile.Name())
	providerImpl := provider.Providers[u.Scheme]
	if providerImpl != nil {
		return fmt.Errorf("unknown schema: %s", target)
	}
	return providerImpl.Upload(ctx, newTarget, tarFile.Name())

}`

Reply:
> and creating/extracting tarball is common task for all providers so I wanted to have that here only

Well, it's not for the `oci` provider (or at least not in the same way).

Reply:
> > and creating/extracting tarball is common task for all providers so I wanted to have that here only
> 
> Well, it's not for the `oci` provider (or at least not in the same way).

I have copied the tarit and extract code from oci only . so ideally this should be same.